[jit] support iteration tuple unpacking

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #22006 [jit] add for in dict support
* #21990 [jit] add for in string support
* **#21985 [jit] support iteration tuple unpacking**
* #21950 [jit] refactor all for in range/tensor tests to be together with other for loop tests
* #21801 [jit] Tree based Iterator infrastructure: for in range/list/tensor/zip/enumerate
* #21949 [jit] more complete tuple assignments

Fixes #20239

Differential Revision: [D15948549](https://our.internmc.facebook.com/intern/diff/D15948549)

wanchaol committed:
[jit] support iteration tuple unpacking

wanchaol committed:
Update on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

wanchaol committed:
rebase with master on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

zdevito commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1314,12 +1314,15 @@ struct to_ir {
-        emitExprsAssign(target_exprs, {sv}, range, n_binders);
+        // if target exprs are more than 1, it means iteration unpacking on LHS
+        // we create Tuple literal to wrap those target exprs for assignments
+        if (target_exprs.size() > 1) {
```
Why is it necessary to create a list here vs just passing more than one target_expr to emitExprsAssign?

wanchaol replied:
Oh it's because `sv` here is the root IterableTree (which create TupleConstruct in `getelem`) and need to call emitTupleAssign to unwrap the tuple using `asTuple`, but the target_exprs here is not a tupleliteral, so `emitExprsAssign` could not match these two together, and we need to create a TupleLiteral wrap here.

zdevito reviewed (approved):
Looks good, one question.

wanchaol committed:
rebase with master on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

driazati commented:
Fixes #20239

wanchaol committed:
rebase with master on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

wanchaol committed:
update comments on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

wanchaol committed:
Update on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head

wanchaol committed:
Update on "[jit] support iteration tuple unpacking"

[jit] support iteration tuple unpacking

gh-metadata: pytorch pytorch 21985 gh/wanchaol/23/head
